waltz-client/src/main/java/com/wepay/waltz/client/internal/network/WaltzNetworkClient.java

```diff
@@ -181,6 +203,33 @@ public void requestHighWaterMark(ReqId reqId) {
         }
     }
 
+    /**
+     * Checks server to storage connectivity.
+     *
+     * Note: There is a possibility that the channel may not be up when this method gets triggered. In that case,
+     * checkServerToStorageConnectivity() gets triggered once the channel is active.
+     *
+     * @return Completable future with the connectivity status to the storage nodes within the cluster.
+     */
+    public CompletableFuture<Optional<Map<String, Boolean>>> checkServerToStorageConnectivity() {
+        ReqId dummyReqId = new ReqId(clientId, 0, 0, seqNumGenerator.incrementAndGet());
+
+        CompletableFuture<Optional<Map<String, Boolean>>> future = checkConnectivityRef.get();
+        if (future != null && isConnectivityTriggered.get()) {
+            return future;
+        }
+
+        if (future == null) {
+            future = new CompletableFuture<>();
+        }
+        if (checkConnectivityRef.compareAndSet(null, future) || !isConnectivityTriggered.get()) {
+            if (sendMessage(new CheckStorageConnectivityRequest(dummyReqId))) {
+                isConnectivityTriggered.set(true);
+            }
+        }
+        return checkConnectivityRef.get();
```

Comment:
You can't return here because there is a race condition. You need loop until `checkConnectivityRef.get()` returns a future or successfully done `checkConnectivityRef.compareAndSet(null, future)`. See `Partition.getHighWaterMark()`. It is a good example of this pattern.